File: simulation/ar_model.py
# -*- coding: utf-8 -*-
"""
Created on Tue May  4 11:24:36 2021
Autoregression model
@author: chenj
"""
import numpy as np
from scipy import linalg

class AR_model:
    def __init__(self, wind_size, dim):
        self.window_len = wind_size
        self.p = dim
        self.hist = []
        self.hist_sin = []
        self.hist_cos = []
        self.real_data_list = []
        self.pred_data_list = []
    
    def update_list(self,input_list,new_value):
        input_list.append(new_value)
        while(len(input_list)>self.window_len):
            input_list.remove(input_list[0])
            
    def predict(self,input_value):
        # use previous value when the history length is smaller than prediction window
        if len(input_value) < self.window_len:
            return input_value[-1]
        try:
            temp = input_value.copy()        
            rxx = np.zeros(self.p+1) #% rxx0--rxx(1), rxx1--rxx(2),..., rxxp--rxx(p+1)    
            for i in range(self.p+1):
                for j in range(self.window_len-i): 
                    rxx[i] = rxx[i] + temp[j]*temp[j+i]                
                rxx[i] = rxx[i]/self.window_len                
            ma = linalg.toeplitz(rxx[:self.p])    
            mb = -1 * rxx[1:]    
            ma_est = np.linalg.solve(ma,mb.T)    
            result = (-1 * ma_est).T @ np.flip(temp[-self.p:])           
            return result
        except:
            return input_value[-1]
    
    def ori_step(self,new_value):
        self.update_list(self.hist_sin,np.sin(new_value*np.pi/180))
        self.update_list(self.hist_cos,np.cos(new_value*np.pi/180))
        pred_sin = self.predict(self.hist_sin)
        pred_cos = self.predict(self.hist_cos)
        pred_value = np.arctan2(pred_sin,pred_cos)*180/np.pi
        self.real_data_list.append(new_value)
        self.pred_data_list.append(pred_value)
        return pred_value
    
    def get_real_list(self):
        return self.real_data_list
    def get_pred_list(self):
        return self.pred_data_list
    
if __name__ == '__main__':
    predWind = 3
    dim = 3
    
    arOriX = AR_model(predWind,dim)
    arOriY = AR_model(predWind,dim)
    with open('./modiTrace.txt') as f:
        msgLen = 64
            
        print("Start playing the trace")
        lines = f.readlines()
        for line in lines:
            if(len(line)>msgLen) :
                print("error")

            # predict the orientation on axis X and Y
            oriX = float(line.split(",")[3])
            predOriX = arOriX.ori_step(oriX)
            oriY = float(line.split(",")[4])
            predOriY = arOriY.ori_step(oriY)
            
            predLine = str(predOriX)+","+str(predOriY)+","+"0"+"\n"

        
        realOriXList = arOriX.get_real_list()
        predOriXList = arOriX.get_pred_list()
        realOriYList = arOriY.get_real_list()
        predOriYList = arOriY.get_pred_list()

        errorOriX = 0.0
        errorOriY = 0.0
        for i in range(len(realOriXList)-1):
            errorOriX += abs((realOriXList[i+1]-predOriXList[i]+180.)%360. - 180.)
            errorOriY += abs((realOriYList[i+1]-predOriYList[i]+180.)%360. - 180.)
            

        errorOriX = errorOriX / (len(realOriXList) - 1)
        errorOriY = errorOriY / (len(realOriYList) - 1)
        print("prediction error on the orientation axis X: ",errorOriX)
        print("prediction error on the orientation axis Y: ",errorOriY)
    

File: server/OfflineServer/ARModel.java
import java.util.*;
import Jama.Matrix;
/**
 * Write a description of ARModel here.
 * 
 * @author (your name) 
 * @version (a version number or a date)
 */
public class ARModel {

    private int predWind;
    private int dim;
    
    private ArrayList<Double> hist;
    private ArrayList<Double> histSin;
    private ArrayList<Double> histCos;
    private ArrayList<Double> realDataList;
    private ArrayList<Double> predDataList;
    
    public ARModel(int window, int p){
        predWind = window;
        dim = p;
        hist = new ArrayList<Double>();
        histSin = new ArrayList<Double>();
        histCos = new ArrayList<Double>();
        realDataList = new ArrayList<Double>();
        predDataList = new ArrayList<Double>();
    }
    
    public ArrayList<Double> getRealList(){
        return realDataList;
    }
    
    public ArrayList<Double> getPredList(){
        return predDataList;
    }
    
    public double posStep(double newValue){
        updateArrayList(hist,newValue);
        double predValue = predict(hist);
        realDataList.add(newValue);
        predDataList.add(predValue);
        return predValue;
    }
    
    public double oriStep(double newValue){
        updateArrayList(histSin,Math.sin(newValue*Math.PI/180));
        updateArrayList(histCos,Math.cos(newValue*Math.PI/180));
        double predSin = predict(histSin);
        double predCos = predict(histCos);
        double predValue = Math.atan2(predSin,predCos)*180/Math.PI;
        realDataList.add(newValue);
        predDataList.add(predValue);
        return predValue;
    }
    
    public void updateArrayList(ArrayList<Double> input, double newValue ){
        input.add(newValue);
        while (input.size() > predWind){
            input.remove(input.get(0));
        }
    }
    
    private Matrix toeplitz(double[] inputArray, int mDim){
        double[][] array = new double[mDim][mDim];
        for(int i=0;i<mDim;i++){
            for(int j=i;j<mDim;j++){
                array[i][j] = inputArray[j-i];
            }
            for(int j=0;j<i;j++){
                array[i][j] = array[j][i];
            }
        }
        return new Matrix(array);
    }

    public double predict(ArrayList<Double> input){
        double result;
        if (input.size() < predWind) {
            result = input.get(input.size()-1);
            return result;
        }
        
        ArrayList<Double> temp = new ArrayList<Double>();
        for (int i=0;i<input.size();i++){
            temp.add(input.get(i));
        }
        try{
            double[] rxx = new double[dim+1];
            for(int j=0;j<dim+1;j++){
                for(int k=0;k<predWind-j;k++){
                    rxx[j] = rxx[j] + temp.get(k).doubleValue()*temp.get(k+j).doubleValue();
                }
                rxx[j] = rxx[j]/predWind;
            }
            Matrix mA = toeplitz(rxx,dim);
            double[][] mbArray = new double[dim][1];
            for(int j=0;j<dim;j++){
                mbArray[j][0] = -1*rxx[1+j];
            }
            Matrix mb = new Matrix(mbArray);
            Matrix maEst = mA.solve(mb);
            double[][] xRevArray = new double[dim][1];
            for(int j=0;j<dim;j++){
                xRevArray[j][0] = temp.get(temp.size()-1-j);
            }
            Matrix xRev = new Matrix(xRevArray);
            //maEst.times(-1).transpose().times(xRev).print(10,10);
            result = maEst.times(-1).transpose().times(xRev).get(0,0);
        } catch(RuntimeException e){
            result = input.get(input.size()-1);
        }
        return result;
    }

    
}
